### ISSUE TITLE
外生变量训练样本长度的问题

### ISSUE THREAD

**Issue by koukouer** [thuml/TimeXer#6]
尊敬的作者：
       您好！感谢开源TimeXer。
       我在读论文的时候注意到外生变量和内生变量的look back len可以被设置为不同，但是在代码中我看到：
self.ex_embedding = DataEmbedding_inverted(configs.seq_len, configs.d_model, configs.embed, configs.freq,configs.dropout)，看上去外生变量的look back len也被设置成了configs.seq_len，与内生变量一致，并且EPF数据集中外生变量都是与内生变量相同长度。
       论文中4.3.1 Mismatch Length中的Increasing Look-back Length实验具体是怎么实现的呀？
       另外，我还想问下，按照这样的处理方式切分样本，理论上每个样本按照configs.seq_len和pred_len切好，还会拥有configs.seq_len长度的外生变量用于辅助预测，我是不是可以把这里的外生变量理解成历史可知变量呢？因为它可能不包含预测周期内产生的外生变量信息。【在时序预测中，外生变量理论上可以分成历史可知变量（长度=内生变量长度）和未来可知变量（长度=内生变量长度+预测步长），我看模型架构里并没有对这两者做区分】

**Comment by wang-lucy** [comment]
您好，可以设置en_seq_len, ex_seq_len两个参数，并在数据加载时dataloader中的seq_len为max(en_seq_len, ex_seq_len)，在模型训练推理时使用内生和外生变量对应长度的序列即可。
理论上模型支持任意长度的外生变量，模型本身并没有限制外生变量的长度一定与内生变量相同，本文为了能与现有的预测模型进行公平比较，在内生和外生变量长度相同的情景下进行实验。我们在文章图11给出了外生变量未来可知情况下的预测效果。

**Comment by lk1983823** [comment]
> 您好，可以设置en_seq_len, ex_seq_len两个参数，并在数据加载时dataloader中的seq_len为max(en_seq_len, ex_seq_len)，在模型训练推理时使用内生和外生变量对应长度的序列即可。 理论上模型支持任意长度的外生变量，模型本身并没有限制外生变量的长度一定与内生变量相同，本文为了能与现有的预测模型进行公平比较，在内生和外生变量长度相同的情景下进行实验。我们在文章图11给出了外生变量未来可知情况下的预测效果。

@wang-lucy 请问在model中具体是在哪里设置en_seq_len, ex_seq_len？

**Comment by fortunerains** [comment]
> 您好，可以设置en_seq_len, ex_seq_len两个参数，并在数据加载时dataloader中的seq_len为max(en_seq_len, ex_seq_len)，在模型训练推理时使用内生和外生变量对应长度的序列即可。 理论上模型支持任意长度的外生变量，模型本身并没有限制外生变量的长度一定与内生变量相同，本文为了能与现有的预测模型进行公平比较，在内生和外生变量长度相同的情景下进行实验。我们在文章图11给出了外生变量未来可知情况下的预测效果。

这个改动大吗 还是得改一下dataloader的部分

**Comment by fortunerains** [comment]
> 您好，可以设置en_seq_len, ex_seq_len两个参数，并在数据加载时dataloader中的seq_len为max(en_seq_len, ex_seq_len)，在模型训练推理时使用内生和外生变量对应长度的序列即可。 理论上模型支持任意长度的外生变量，模型本身并没有限制外生变量的长度一定与内生变量相同，本文为了能与现有的预测模型进行公平比较，在内生和外生变量长度相同的情景下进行实验。我们在文章图11给出了外生变量未来可知情况下的预测效果。

这两个参数代码中并无

**Comment by linbingkong** [comment]
> 您好，可以设置en_seq_len, ex_seq_len两个参数，并在数据加载时dataloader中的seq_len为max(en_seq_len, ex_seq_len)，在模型训练推理时使用内生和外生变量对应长度的序列即可。 理论上模型支持任意长度的外生变量，模型本身并没有限制外生变量的长度一定与内生变量相同，本文为了能与现有的预测模型进行公平比较，在内生和外生变量长度相同的情景下进行实验。我们在文章图11给出了外生变量未来可知情况下的预测效果。

该种方法是否合理？seq_len为max(en_seq_len, ex_seq_len)，因为在该种情况下 ex_seq_len肯定大于en_seq_len,因此就相当于seq_len=ex_seq_len,但该做法本质上就只拉长了seq_len,pred_len的部分也会同步向后移，并没有将pred_len 外甥变量的部分数据作为输入

**Comment by SamuelWong77** [comment]
> > 您好，可以设置en_seq_len, ex_seq_len两个参数，并在数据加载时dataloader中的seq_len为max(en_seq_len, ex_seq_len)，在模型训练推理时使用内生和外生变量对应长度的序列即可。 理论上模型支持任意长度的外生变量，模型本身并没有限制外生变量的长度一定与内生变量相同，本文为了能与现有的预测模型进行公平比较，在内生和外生变量长度相同的情景下进行实验。我们在文章图11给出了外生变量未来可知情况下的预测效果。
> 
> 该种方法是否合理？seq_len为max(en_seq_len, ex_seq_len)，因为在该种情况下 ex_seq_len肯定大于en_seq_len,因此就相当于seq_len=ex_seq_len,但该做法本质上就只拉长了seq_len,pred_len的部分也会同步向后移，并没有将pred_len 外甥变量的部分数据作为输入

模式是encoder only吧

**Comment by linbingkong** [comment]
> > > 您好，可以设置en_seq_len, ex_seq_len两个参数，并在数据加载时dataloader中的seq_len为max(en_seq_len, ex_seq_len)，在模型训练推理时使用内生和外生变量对应长度的序列即可。 理论上模型支持任意长度的外生变量，模型本身并没有限制外生变量的长度一定与内生变量相同，本文为了能与现有的预测模型进行公平比较，在内生和外生变量长度相同的情景下进行实验。我们在文章图11给出了外生变量未来可知情况下的预测效果。
> > 
> > 
> > 该种方法是否合理？seq_len为max(en_seq_len, ex_seq_len)，因为在该种情况下 ex_seq_len肯定大于en_seq_len,因此就相当于seq_len=ex_seq_len,但该做法本质上就只拉长了seq_len,pred_len的部分也会同步向后移，并没有将pred_len 外甥变量的部分数据作为输入
> 
> 模式是encoder only吧

dec_inp = torch.zeros_like(batch_y[:, -self.args.pred_len:, :]).float()
dec_inp = torch.cat([batch_y[:, :self.args.label_len, :], dec_inp], dim=1).float().to(self.device)
padding的时候也会把pred_len部分的外生变量置为0

**Comment by SamuelWong77** [comment]
> > > > 您好，可以设置en_seq_len, ex_seq_len两个参数，并在数据加载时dataloader中的seq_len为max(en_seq_len, ex_seq_len)，在模型训练推理时使用内生和外生变量对应长度的序列即可。 理论上模型支持任意长度的外生变量，模型本身并没有限制外生变量的长度一定与内生变量相同，本文为了能与现有的预测模型进行公平比较，在内生和外生变量长度相同的情景下进行实验。我们在文章图11给出了外生变量未来可知情况下的预测效果。
> > > 
> > > 
> > > 该种方法是否合理？seq_len为max(en_seq_len, ex_seq_len)，因为在该种情况下 ex_seq_len肯定大于en_seq_len,因此就相当于seq_len=ex_seq_len,但该做法本质上就只拉长了seq_len,pred_len的部分也会同步向后移，并没有将pred_len 外甥变量的部分数据作为输入
> > 
> > 
> > 模式是encoder only吧
> 
> dec_inp = torch.zeros_like(batch_y[:, -self.args.pred_len:, :]).float() dec_inp = torch.cat([batch_y[:, :self.args.label_len, :], dec_inp], dim=1).float().to(self.device) padding的时候也会把pred_len部分的外生变量置为0

dec_inp不做模型输入outputs = self.model(enc_inp, batch_x_mark)[0]。
 然后可以分别对内生变量的pred len填充0，外生变量则不padding，然后二者在feature维度拼接恢复通道数，注意外生变量在最后一个维度。

**Comment by linbingkong** [comment]
> > > > > 您好，可以设置en_seq_len, ex_seq_len两个参数，并在数据加载时dataloader中的seq_len为max(en_seq_len, ex_seq_len)，在模型训练推理时使用内生和外生变量对应长度的序列即可。 理论上模型支持任意长度的外生变量，模型本身并没有限制外生变量的长度一定与内生变量相同，本文为了能与现有的预测模型进行公平比较，在内生和外生变量长度相同的情景下进行实验。我们在文章图11给出了外生变量未来可知情况下的预测效果。
> > > > 
> > > > 
> > > > 该种方法是否合理？seq_len为max(en_seq_len, ex_seq_len)，因为在该种情况下 ex_seq_len肯定大于en_seq_len,因此就相当于seq_len=ex_seq_len,但该做法本质上就只拉长了seq_len,pred_len的部分也会同步向后移，并没有将pred_len 外甥变量的部分数据作为输入
> > > 
> > > 
> > > 模式是encoder only吧
> > 
> > 
> > dec_inp = torch.zeros_like(batch_y[:, -self.args.pred_len:, :]).float() dec_inp = torch.cat([batch_y[:, :self.args.label_len, :], dec_inp], dim=1).float().to(self.device) padding的时候也会把pred_len部分的外生变量置为0
> 
> dec_inp不做模型输入outputs = self.model(enc_inp, batch_x_mark)[0]。 然后可以分别对内生变量的pred len填充0，外生变量则不padding，然后二者在feature维度拼接恢复通道数，注意外生变量在最后一个维度。

明白了
# encoder input
enc_inp = batch_x[:, -self.args.pred_len:, :].clone()
# 只将标签列置为0，比如标签是最后一个特征列
label_col = -1  # 或者指定具体的列索引，如 0 或 2
enc_inp[:, :, label_col] = 0
enc_inp = torch.cat([enc_inp[:, :self.args.label_len, :], enc_inp], dim=1).float().to(self.device)

outputs = self.model(enc_inp, batch_x_mark, dec_inp, batch_y_mark)

**Comment by linbingkong** [comment]
改起来还是非常麻烦，包括模型的某些参数也要修改，也不太确定对不对。希望作者能提供一版给大家参考

**Comment by SamuelWong77** [comment]
> 您好，可以设置en_seq_len, ex_seq_len两个参数，并在数据加载时dataloader中的seq_len为max(en_seq_len, ex_seq_len)，在模型训练推理时使用内生和外生变量对应长度的序列即可。 理论上模型支持任意长度的外生变量，模型本身并没有限制外生变量的长度一定与内生变量相同，本文为了能与现有的预测模型进行公平比较，在内生和外生变量长度相同的情景下进行实验。我们在文章图11给出了外生变量未来可知情况下的预测效果。

作者您好 请问模型支持外生内生长度不同  那么有2个问题：
（1）内生外生look back不同时，时间戳取长的吗，而短视的变量是否要padding为0？
（2）pred时如果外生预测作为模型输入，内生部分是否要padding为0？
